[FLINK-17239][docs] Streaming Analytics tutorial (#11843)
apache/flink: docs/tutorials/streaming_analytics.md

implementations.

### Working with Event Time

By default, Flink will use processing time. To change this, you can set the Time Characteristic:

{% highlight java %}
final StreamExecutionEnvironment env =
    StreamExecutionEnvironment.getExecutionEnvironment();
env.setStreamTimeCharacteristic(TimeCharacteristic.EventTime);
{% endhighlight %}

If you want to use event time, you will also need to supply a Timestamp Extractor and Watermark
Generator that Flink will use to track the progress of event time. This will be covered in the
section below on [Working with Watermarks]({{ site.baseurl }}{% link
tutorials/streaming_analytics.md %}#working-with-watermarks), but first we should explain what
watermarks are.

### Watermarks

Let's work through a simple example that will show why watermarks are needed, and how they work.

In this example you have a stream of timestamped events that arrive somewhat out of order, as shown
below. The numbers shown are timestamps that indicate when these events actually occurred. The first
event to arrive happened at time 4, and it is followed by an event that happened earlier, at time 2,
and so on:

>     ··· 23 19 22 24 21 14 17 13 12 15 9 11 7 2 4 →

Reviewer: I think, I would prefer to have this as a figure, or at least increase the font size and center it?

Author: I made it more presentable. (But I confess that the html I used isn't very pretty.)

Reviewer: yes, I would agree on that ;)

let me have a look at improving the html a bit...

Reviewer: this seems to look nice (I'll apply it to your PR unless you object):
```
<div class="text-center" style="font-size: x-large; word-spacing: 0.5em; margin: 1em 0em;">
··· 23 19 22 24 21 14 17 13 12 15 9 11 7 2 4 →
</div>
```

4, and then an event from time 2. Will an event with a timestamp less than 2 ever arrive? Maybe.
Maybe not. You could wait forever and never see a 1.

*Eventually you have to be courageous and emit the 2 as the start of the sorted stream.*

(3) What you need then is some sort of policy that defines when, for any given timestamped event, to
stop waiting for the arrival of earlier events.

*This is precisely what watermarks do* — they define when to stop waiting for earlier events.

Event time processing in Flink depends on *watermark generators* that insert special timestamped
elements into the stream, called *watermarks*. A watermark for time _t_ is an assertion that the
stream is (probably) now complete up through time _t_.

When should this stream sorter stop waiting, and push out the 2 to start the sorted stream? When a
watermark arrives with a timestamp of 2, or greater.

(4) You might imagine different policies for deciding how to generate watermarks.

Each event arrives after some delay, and these delays vary, so some events are delayed more than
others. One simple approach is to assume that these delays are bounded by some maximum delay. Flink

> refers to this strategy as *bounded-out-of-orderness* watermarking. It's easy to imagine more

Reviewer: ```suggestion
refers to this strategy as *bounded-out-of-orderness* watermarking. It is easy to imagine more
```

### Latency vs. Completeness

Another way to think about watermarks is that they give you, the developer of a streaming
application, control over the tradeoff between latency and completeness. Unlike in batch processing,
where one has the luxury of being able to have complete knowledge of the input before producing any
results, with streaming you must eventually stop waiting to see more of the input, and produce some
sort of result.

You can either configure your watermarking aggressively, with a short bounded delay, and thereby
take the risk of producing results with rather incomplete knowledge of the input -- i.e., a possibly
wrong result, produced quickly. Or you can wait longer, and produce results that take advantage of
having more complete knowledge of the input stream(s).

It is also possible to implement hybrid solutions that produce initial results quickly, and then
supply updates to those results as additional (late) data is processed. This is a good approach for
some applications.

### Lateness

Lateness is defined relative to the watermarks. A `Watermark(t)` asserts that the stream is complete

> up through time _t_; any event following this watermark whose timestamp is &le; _t_ is late.

Reviewer: I think, it should be this, shouldn't it?
```suggestion
up thru time _t_; any event following this watermark whose timestamp is &le; _t_ is late.
```

Author: "Through is the only formally accepted spelling of the word. Thru is an alternate spelling that should be used only in informal writing or when referring to drive-throughs."

Reviewer: oh, I guess, then the other "thru" that I saw in a previous PR may need changing...(I was wondering there, looked it up quickly, and then proposed it here as a result)

### Working with Watermarks

In order to perform event-time-based event processing, Flink needs to know the time associated with
each event, and it also needs the stream to include watermarks.

The Taxi data sources used in the hands-on exercises take care of these details for you. But in your
own applications you will have to take care of this yourself, which is usually done by implementing
a class that extracts the timestamps from the events, and generates watermarks on demand. The
easiest way to do this is by extending the `BoundedOutOfOrdernessTimestampExtractor`:

{% highlight java %}
DataStream<Event> stream = ...

DataStream<Event> withTimestampsAndWatermarks =
    stream.assignTimestampsAndWatermarks(new TimestampsAndWatermarks(Time.seconds(10)));

public static class TimestampsAndWatermarks

>         extends BoundedOutOfOrdernessTimestampExtractor<MyEvent> {

Reviewer: ```suggestion
        extends BoundedOutOfOrdernessTimestampExtractor<Event> {
```

    public TimestampsAndWatermarks(Time t) {
        super(t);
    }

    @Override
    public long extractTimestamp(Event event) {
        return event.timestamp;
    }
}
{% endhighlight %}

Note that the constructor for `BoundedOutOfOrdernessTimestampExtractor` takes a parameter which
specifies the maximum expected out-of-orderness (10 seconds, in this example).

{% top %}

## Windows

Flink features very expressive window semantics.

In this section you will learn:

* how windows are used to compute aggregates on unbounded streams,
* which types of windows Flink supports, and

> * how to implement a DataStream program with a window aggregation

Reviewer: maybe rather this?
```suggestion
* how to implement a DataStream program with a windowed aggregation
```

### Introduction

> It's natural when doing stream processing to want to compute aggregated analytics on bounded subsets

Reviewer: ```suggestion
It is natural when doing stream processing to want to compute aggregated analytics on bounded subsets
```

    .reduce|aggregate|process(<window function>)
{% endhighlight %}

### Window Assigners

Flink has several built-in types of window assigners, which are illustrated below:

<img src="{{ site.baseurl }}/fig/window-assigners.svg" alt="Window assigners" class="center" width="80%" />

Some examples of what these window assigners might be used for, and how to specify them:

* Tumbling time windows
  * _page views per minute_
  * `TumblingEventTimeWindows.of(Time.minutes(1))`
* Sliding time windows
  * _page views per minute computed every 10 seconds_
  * `SlidingEventTimeWindows.of(Time.minutes(1), Time.seconds(10))`
* Session windows 
  * _page views per session, where sessions are defined by a gap of at least 30 minutes between sessions_
  * `EventTimeSessionWindows.withGap(Time.minutes(30))`

Durations can be specified using one of `Time.milliseconds(n)`, `Time.seconds(n)`, `Time.minutes(n)`, `Time.hours(n)`, and `Time.days(n)`.

The time-based window assigners (including session windows) come in both event time and processing
time flavors. There are significant tradeoffs between these two types of time windows. With
processing time windowing you have to accept these limitations:

> * can not process historic data,

Reviewer: ```suggestion
* can not correctly process historic data,
```

1. or with a combination of the two, wherein the pre-aggregated results of a `ReduceFunction` or an `AggregateFunction` are supplied to a `ProcessWindowFunction` when the window is triggered.

Here are examples of approaches 1 and 3. Each implementation finds the peak value from each sensor
in 1 minute event time windows, and producing a stream of Tuples containing `(key,
end-of-window-timestamp, max_value)`.

#### ProcessWindowFunction Example

{% highlight java %}
DataStream<SensorReading> input = ...

input
    .keyBy(x -> x.key)
    .window(TumblingEventTimeWindows.of(Time.minutes(1)))
    .process(new MyWastefulMax());

public static class MyWastefulMax extends ProcessWindowFunction<
        SensorReading,                  // input type
        Tuple3<String, Long, Integer>,  // output type
        String,                         // key type
        TimeWindow> {                   // window type
    
    @Override
    public void process(
            String key,
            Context context, 
            Iterable<SensorReading> events,
            Collector<Tuple3<String, Long, Integer>> out) {

        int max = 0;
        for (SensorReading event : events) {

>             if (event.value > max) max = event.value;

Reviewer: These on-liners are often rather confusing / sources of errors. I actually think, I have seen this among the IntelliJ warnings as well for doing Math.max() on your own, so maybe replace by:
```suggestion
            max = Math.max(event.value, max);
```

        }

>         out.collect(new Tuple3<>(key, context.window().getEnd(), max));

Reviewer: ```suggestion
        out.collect(Tuple3.of(key, context.window().getEnd(), max));
```

    }
}
{% endhighlight %}

A couple of things to note in this implementation:

* All of the events assigned to the window have to be buffered in keyed Flink state until the window
  is triggered. This is potentially quite expensive.

> * Our ProcessWindowFunction is being passed a Context object from which contains information about

Reviewer: ```suggestion
* Our `ProcessWindowFunction` is being passed a `Context` object from which contains information about
```

  the window. Its interface looks like this:

{% highlight java %}
public abstract class Context implements java.io.Serializable {
    public abstract W window();
    
    public abstract long currentProcessingTime();
    public abstract long currentWatermark();

    public abstract KeyedStateStore windowState();
    public abstract KeyedStateStore globalState();
}
{% endhighlight %}

> `windowState` and `globalState` are places where you can store per-key, per-window, or global
> per-key information. This might be useful, for example, if you want to record something about the

Reviewer: I often find this distinction (naming) a bit confusing. How about this to make it clear(er)?
```suggestion
`windowState` and `globalState` are places where you can store per-key, per-window, or global
per-key information for all windows of that key. This might be useful, for example, if you want to
record something about the
```

current window and use that when processing a subsequent window.

#### Incremental Aggregation Example

{% highlight java %}
DataStream<SensorReading> input = ...

input
    .keyBy(x -> x.key)
    .window(TumblingEventTimeWindows.of(Time.minutes(1)))
    .reduce(new MyReducingMax(), new MyWindowFunction());

private static class MyReducingMax implements ReduceFunction<SensorReading> {
    public SensorReading reduce(SensorReading r1, SensorReading r2) {
        return r1.value() > r2.value() ? r1 : r2;
    }
}

private static class MyWindowFunction extends ProcessWindowFunction<
    SensorReading, Tuple3<String, Long, SensorReading>, String, TimeWindow> {

    @Override
    public void process(
            String key,
            Context context,
            Iterable<SensorReading> maxReading,
            Collector<Tuple3<String, Long, SensorReading>> out) {

        SensorReading max = maxReading.iterator().next();

>         out.collect(new Tuple3<String, Long, SensorReading>(key, context.window().getEnd(), max));

Reviewer: ```suggestion
        out.collect(Tuple3.of(key, context.window().getEnd(), max));
```

    }
}
{% endhighlight %}

> This implementation uses a more robust `KeySelector`. Notice also that the `Iterable<SensorReading>`

Reviewer: The `KeySelector` is actually the same...

Author: Oops.

will contain exactly one reading -- the pre-aggregated maximum computed by `MyReducingMax`.

### Late Events

By default, when using event time windows, late events are dropped. There are two optional parts of
the window API that give you more control over this.

You can arrange for the events that would be dropped to be collected to an alternate output stream
instead, using a mechanism called [Side Outputs]({{ site.baseurl }}{% link tutorials/event_driven.md

> %}#side-outputs). Here's an example of what that might look like:

Reviewer: ```suggestion
%}#side-outputs). Here is an example of what that might look like:
```

{% highlight java %}
OutputTag<Event> lateTag = new OutputTag<Event>("late"){};

SingleOutputStreamOperator<Event> result = stream.
    .keyBy(...)
    .window(...)
    .sideOutputLateData(lateTag)
    .process(...);
  
DataStream<Event> lateStream = result.getSideOutput(lateTag);
{% endhighlight %}

You can also specify an interval of _allowed lateness_ during which the late events will continue to
be assigned to the appropriate window(s) (whose state will have been retained). By default each late

> event will cause a late firing of the window function.

Reviewer: Also explain what a late firing means?

Author: done

sent to the side output).

For example:

{% highlight java %}
stream.
    .keyBy(...)
    .window(...)
    .allowedLateness(Time.seconds(10))
    .process(...);
{% endhighlight %}

When the allowed lateness is greater than zero, only those events that are so late that they would
be dropped are sent to the side output (if it has been configured).

### Surprises

Some aspects of Flink's windowing API may not behave in the way you would expect. Based on
frequently asked questions on the [flink-user mailing
list](https://flink.apache.org/community.html#mailing-lists) and elsewhere, here are some facts
about windows that may surprise you.

#### Sliding Windows Make Copies

Sliding window assigners can create lots of window objects, and will copy each event into every
relevant window. For example, if you have sliding windows every 15 minutes that are 24-hours in
length, each event will be copied into 4 * 24 = 96 windows.

#### Time Windows are Aligned to the Epoch

Just because you are using hour-long processing-time windows and start your application running at
12:05 does not mean that the first window will close at 1:05. The first window will be 55 minutes

> long and close at 1:00.

Reviewer: Should we mention the ability to set the offset here as a workaround?

Author: Sure. Done.

#### Windows Can Follow Windows

For example, it works to do this:

{% highlight java %}
stream
    .keyBy(t -> t.key)
    .timeWindow(<time specification>)
    .reduce(<reduce function>)
    .timeWindowAll(<same time specification>)
    .reduce(<same reduce function>)
{% endhighlight %}

You might expect Flink's runtime to be smart enough to do this parallel pre-aggregation for you

> (provided you are using a ReduceFunction or AggregateFunction), but it's not.

Reviewer: Add requirements on the time specifications to be compatible?

Author: done

#### No Results for Empty TimeWindows

Windows are only created when events are assigned to them. So if there are no events in a given time
frame, no results will be reported.

#### Late Events Can Cause Late Merges

Session windows are based on an abstraction of windows that can _merge_. Each element is initially
assigned to a new window, after which windows are merged whenever the gap between them is small
enough. In this way, a late event can bridge the gap separating two previously separate sessions,
producing a late merge.

{% top %}

## Hands-on

The hands-on exercise that goes with this section is the [Hourly Tips

> Exercise](https://github.com/apache/flink-training/tree/master/hourly-tips).

Reviewer: same as in the previous PR: is it possible to link to the right branch?

Author: Already done.
